docarray/array/array.py

```diff
@@ -87,24 +91,118 @@ class Image(BaseDocument):
 
     def __init__(
         self,
-        docs: Iterable[BaseDocument] = list(),
+        docs: Optional[Iterable[BaseDocument]] = None,
         tensor_type: Type['AbstractTensor'] = NdArray,
     ):
-        self._data = [doc_ for doc_ in docs]
+        self._data = list(docs) if docs is not None else []
         self.tensor_type = tensor_type
 
     def __len__(self):
         return len(self._data)
 
-    def __getitem__(self, item):
+    def __getitem__(self: T, item: IndexType) -> Union[BaseDocument, T]:
+        item = self._normalize_index_item(item)
+
         if type(item) == slice:
             return self.__class__(self._data[item])
-        else:
+
+        if isinstance(item, int):
             return self._data[item]
 
+        if item is None:
+            return self
+
+        # _normalize_index_item() guarantees the line below is correct
+        head = item[0]  # type: ignore
+        if isinstance(head, bool):
+            return self._get_from_mask(item)
+        elif isinstance(head, int):
+            return self._get_from_indices(item)
+        else:
+            raise TypeError(f'Invalid type {type(head)} for indexing')
+
+    def __setitem__(self: T, key: IndexType, value: Union[T, BaseDocument]):
+        key = self._normalize_index_item(key)
+
+        if isinstance(key, (int, slice)):
+            self._data[key] = value
+        else:
+            # _normalize_index_item() guarantees the line below is correct
+            head = key[0]  # type: ignore
+            if isinstance(head, bool):
+                return self._set_by_mask(key, value)
+            elif isinstance(head, int):
+                return self._set_by_indices(key, value)
+            else:
+                raise TypeError(f'Invalid type {type(head)} for indexing')
+
     def __iter__(self):
         return iter(self._data)
 
+    @staticmethod
+    def _normalize_index_item(
+        item: Any,
+    ) -> Union[int, slice, Iterable[int], Iterable[bool], None]:
+        if item is None or isinstance(item, (int, slice, tuple, list)):
+            return item
+
+        index_has_getitem = hasattr(item, '__getitem__')
+        is_valid_bulk_index = index_has_getitem and isinstance(item, Iterable)
+        if not is_valid_bulk_index:
+            raise ValueError(f'Invalid index type {type(item)}')
+
+        if isinstance(item, np.ndarray) and (
+            item.dtype == np.bool or np.issubdtype(item.dtype, np.integer)
+        ):
+            return item.tolist()
+
+        try:
+            import torch  # noqa: F401
```

Comment:
lets do this import at the tope of the file. Torch import can be slow.

we do this a couple of other time in the codebase it looks like this

```python
try:
    import torch
    torch_imported = True
except ImportError:
    torch_imported = False
```

Reply:
then I guess we should do this once globally and import it everywhere else

Reply:
but where globaly ?